ARROW-13576: [C++] Replace ExecNode::InputReceived with ::MakeTask

aocsa opened the pull request:
WIP

felipeblazing commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -99,11 +105,35 @@ class FilterNode : public ExecNode {
-    if (ErrorIfNotOk(maybe_filtered.status())) return;
+    ARROW_LOG(DEBUG) << "FilterNode: >> input";
+
+    auto executor = plan()->exec_context()->executor();
```
Is the purpose of the logic here to allow the FilterNode to still be used in the old way while allowing it schedule a task as well?

aocsa replied:
Well executor is optional so I think will be maintained to run without executor.

lidavidm replied:
While that's currently true, I think we want to require an executor in the future (even if the executor is "block and do the work on the calling thread"). See https://github.com/apache/arrow/pull/10845#issuecomment-904986082 and ARROW-13741.

felipeblazing commented on cpp/src/arrow/compute/exec/project_node.cc:
```diff
@@ -91,6 +91,7 @@ class ProjectNode : public ExecNode {
 
   void InputReceived(ExecNode* input, ExecBatch batch) override {
     DCHECK_EQ(input, inputs_[0]);
+    ARROW_LOG(DEBUG) << "ProjectNode: >> project input >>" + batch.ToString();
```
I take it this is debugging code specific to your test and this will be removed?

aocsa replied:
sure

felipeblazing commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -99,11 +105,35 @@ class FilterNode : public ExecNode {
+
+    auto executor = plan()->exec_context()->executor();
+    if (executor) {
+      auto maybe_future = executor->Submit([this, batch] {
```
Whatever we submit to the executor must be more than just a lambda if it is to encode information that the scheduler needs to use. A cost function and context about what query this belongs to and other metadata might want to be part of what the task includes.

felipeblazing replied:
I am not saying that we need to include those things now but it would be good to see if there is a way to send this information with the task

aocsa replied:
done

felipeblazing commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -99,11 +105,35 @@ class FilterNode : public ExecNode {
+        if (ErrorIfNotOk(maybe_filtered.status())) Status::Invalid("not ok");
+
+        maybe_filtered->guarantee = batch.guarantee;
+        this->outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
```
This feels like you would struggle to assure thread affinity as you went from task to task here.

felipeblazing replied:
The other proposal was to have the task build up and input_received receive a task with the first node in some plan receiving a task that is something akin to, my task is giving you this data.

lidavidm commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -113,7 +131,22 @@ class FilterNode : public ExecNode {
     DCHECK_EQ(input, inputs_[0]);
+    auto executor = plan()->exec_context()->executor();
+
+    for (auto&& task : tasks_) {
```
InputFinished (despite its name) can be called before InputReceived (it's less of an "input is done" and more of "this is how many inputs to expect"). So this won't necessarily work; you'll want something like the AtomicCounter pattern used in other exec nodes.

aocsa replied:
Thanks  @lidavidm  I am using AtomicCounters and there is not more race condition issues

westonpace commented on cpp/src/arrow/compute/exec/exec_plan.cc:
```diff
@@ -283,6 +285,19 @@ bool ExecNode::ErrorIfNotOk(Status status) {
+  auto executor = plan()->exec_context()->executor();
+  auto maybe_future = executor->Submit(std::move(task));
+  if (!maybe_future.ok()) {
+    outputs_[0]->ErrorReceived(this, maybe_future.status());
```
Nit: It doesn't matter today but if this is a general method we may want to loop through all outputs.

westonpace commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -98,12 +97,25 @@ class FilterNode : public ExecNode {
+      outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+      return Status::OK();
+    };
+    if (this->has_executor()) {
+      DCHECK_OK(this->SubmitTask(task));
+    } else {
+      DCHECK_OK(task());
+    }
+    if (batch_count_.Increment()) {
+      task_group_.WaitForTasksToFinish().AddCallback(
+          [this](const Status& status) { this->finished_.MarkFinished(status); });
+    }
```
Could this be part of SubmitTask so that this if/else loop doesn't have to belong in every node?

westonpace commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -98,12 +97,25 @@ class FilterNode : public ExecNode {
-
-    maybe_filtered->guarantee = batch.guarantee;
-    outputs_[0]->InputReceived(this, maybe_filtered.MoveValueUnsafe());
+    if (finished_.is_finished()) {
+      return;
+    }
```
Could this be part of SubmitTask?

aocsa replied:
Sure

westonpace commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -114,6 +126,10 @@ class FilterNode : public ExecNode {
   void InputFinished(ExecNode* input, int total_batches) override {
     DCHECK_EQ(input, inputs_[0]);
     outputs_[0]->InputFinished(this, total_batches);
+    if (batch_count_.SetTotal(total_batches)) {
+      task_group_.WaitForTasksToFinish().AddCallback(
+          [this](const Status& status) { this->finished_.MarkFinished(status); });
+    }
```
This also feels like code that will have to be repeated in every node.

westonpace commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -127,21 +143,25 @@ class FilterNode : public ExecNode {
 
-  void StopProducing() override { inputs_[0]->StopProducing(this); }
+  void StopProducing() override {
+    if (batch_count_.Cancel()) {
+      task_group_.WaitForTasksToFinish().AddCallback(
+          [this](const Status& status) { this->finished_.MarkFinished(status); });
+    }
```
This code feels like it will need to be repeated

westonpace commented on cpp/src/arrow/util/task_scheduler.h:
```diff
@@ -0,0 +1,81 @@
+namespace arrow {
+namespace internal {
+
+class ARROW_EXPORT TaskScheduler : public Executor {
+ public:
+  static Result<std::shared_ptr<TaskScheduler>> Make(int threads);
+  static Result<std::shared_ptr<TaskScheduler>> MakeEternal(int threads);
+
+  ~TaskScheduler() override;
+
+  int GetCapacity() override;
+
+  bool OwnsThisThread() override;
+
+  std::shared_ptr<ThreadPool> pool() { return thread_pool_; }
+
+ protected:
+  friend ARROW_EXPORT TaskScheduler* GetCpuTaskScheduler();
+
+  TaskScheduler();
+
+  Status SpawnReal(TaskHints hints, FnOnce<void()> task, StopToken,
+                   StopCallback&&) override;
+
+  static std::shared_ptr<TaskScheduler> MakeCpuTaskScheduler();
+
+ private:
+  std::shared_ptr<ThreadPool> thread_pool_;
+  // std::queue<std::unique_ptr<Task>> task_queue_;
+  // std::atomic<int> active_tasks_counter_;
+  //
+};
```
I'm not really sure what the purpose is here.  I thought "a scheduler has many executors" but these seems like "a scheduler is an executor".  I don't think the scheduler should need `MakeEternal` and there shouldn't be `MakeCpuTaskScheduler`.

Maybe there is a `MakeDefaultScheduler` which has only a single executor (the CPU executor).  The scheduler shouldn't worry about `OwnsThisThread` or `GetCapacity`.

felipeblazing replied:
I agree that a scheduler should be able to have multiple executors. It also will eventually have more awareness of the entire system than an executor might have.

aocsa replied:
Agree. TaskScheduler should have multiple executors. I think it can be part of follow-up PR after discussing the next steps.

westonpace reviewed (commented):
Looks like a reasonable start.  I'm not sure I fully understand the task scheduler right now and it would be nice if we could avoid duplication between the nodes on some of the task management work.

aocsa commented:
I update this PR and I think is ready to review.  I  was struggling trying to work fine with `AsyncTaskGroup` but it seems this util has an issue when concurrent thread are trying to know when `all_task` are `done`.  See here: https://gist.github.com/aocsa/c4e66e9061d1dc9f1a0e8253ffad2be3

So now I am using TaskGroup class in async mode with a `stop_token` to cancel tasks when `ExecNode::StopProducing`  is requested. cc @felipeblazing, @westonpace

lidavidm commented on cpp/src/arrow/util/thread_pool.h:
```diff
@@ -341,6 +341,8 @@ class ARROW_EXPORT ThreadPool : public Executor {
   Status Shutdown(bool wait = true);
 
+  Status SpawnReal(TaskHints hints, FnOnce<void()> task, StopToken,
+                   StopCallback&&) override;
```
Do we still need to make this public?

lidavidm commented on cpp/src/arrow/compute/exec/exec_plan.cc:
```diff
@@ -283,6 +290,35 @@ bool ExecNode::ErrorIfNotOk(Status status) {
+      this->stop_source_.RequestStop();
+    }
+    task_group_->FinishAsync().AddCallback([this](const Status& status) {
+      if (!this->finished_.is_finished()) this->finished_.MarkFinished(status);
```
This is racy/this is a TOC-TOU error.

lidavidm replied:
So long as this is only called after checking AtomicCounter, is the check here necessary?

aocsa replied:
Based on my test and profiling test, yes this is necessary. The first call to this functions could when `batch_count_` reach the total_batches, but at the same time ExecNode::StopProducing signal could happens. So only once it's neccesary to mark as Finished  `finished_`.

lidavidm replied:
Hmm. I think we have a race, then. The point of AtomicCounter is so that we only ever make one call to MarkFinished. I also just noticed that batch_count_.Increment() needs to go inside the task, not outside it.

lidavidm commented on cpp/src/arrow/compute/exec/exec_plan.h:
```diff
@@ -221,6 +228,18 @@ class ARROW_EXPORT ExecNode {
   std::string ToString() const;
 
+  /// \brief Is an executor available?
+  bool has_executor() { return plan()->exec_context()->executor() != nullptr; }
```
I think all three new methods should be protected (also has_executor can be const, no?)

aocsa replied:
For sure

lidavidm commented on cpp/src/arrow/compute/exec/exec_plan.h:
```diff
@@ -241,6 +260,15 @@ class ARROW_EXPORT ExecNode {
+  // Counter for the number of batches received
+  AtomicCounter batch_count_;
+  // Future to sync finished
+  Future<> finished_ = Future<>::MakeFinished();
```
No need to initialize here if we're also initializing in the constructor.

lidavidm commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -127,21 +131,24 @@ class FilterNode : public ExecNode {
+  }
 
-  Future<> finished() override { return inputs_[0]->finished(); }
+  Future<> finished() override { return finished_; }
```
Maybe this should be the default implementation in the base class now? (Though I guess for this PR that requires more refactoring.)

lidavidm replied:
(And in general, if we're codifying patterns like `AtomicCounter batch_count_`, I think more of these boilerplate methods could be implemented in the base class instead, once we're OK with the overall design in this PR.)

lidavidm reviewed (commented):
Some implementation-level comments. Thanks for finding the error in AsyncTaskGroup; we should figure out what's going on there as well.

lidavidm commented:
Ah, for AsyncTaskGroup, I think what's happening is that it allows adding tasks even after WaitForTaskToFinish, but it's assumed that this will only be done by another task. But for this PR, presumably that's not the case - we're adding tasks externally so that invariant may be violated, and so running_tasks_ may hit 0, the future may be marked finished, and then a concurrent call may add another task. So the semantics of AsyncTaskGroup don't quite fit what we want here.

westonpace commented:
FYI, in the new dataset writer node PR, I've renamed `WaitForTaskToFinish` to `End` so it's clear that it should only be called once.  I've also added `OnFinished` which returns a future and can be accessed early.  So you can call `OnFinished` as much as you want (e.g. before you add any tasks even) but you only call `End` once.  I think this clears up the confusion.

Check it out here: https://github.com/apache/arrow/pull/11017/files

If you want to use this I can pull out the changes to AsyncTaskGroup and submit them as a small PR on their own today.  Otherways, it will probably be some ways off before 11017 merges.

westonpace reviewed (commented):
I like the idea of combining similar logic in one place but I don't think this implementation is not going to make sense for all exec nodes.  Rather than write one huge comment I drafted up a bare sketch of an idea here: https://docs.google.com/document/d/1aJWvU-9fxnl8eE6Iw-IebVVs8wxkZ30f1rNpSBOBgmY/edit#heading=h.qv2yr7aeayxs

I didn't spend too much time on it so feel free to reject it out of the box.  However, I think we will want to either create various ExecNode subclasses or create an "exec node runner" concept.  This way we keep the base ExecNode API pure.  This allows nodes that need to have a rather custom implementation to keep their custom implementation and nodes that can share common logic can do so.

aocsa commented:
I refactor this PR based of your feedbacks. 
Main change:
- Refactor using composition to create two implementations `MakeSimpleSyncRunner` and `MakeSimpleParallelRunner` for the use in filer and project node. 
- Fixed some race condition
- Use of AsyncTaskGroup again.
cc @westonpace, @felipeblazing, @lidavidm

lidavidm commented on cpp/src/arrow/util/future.h:
```diff
@@ -840,6 +840,18 @@ inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s
+/// Unlike AllComplete this Future will not complete immediately when a
+/// failure occurs.  It will wait until all futures have finished.
+ARROW_EXPORT
+Future<> AllCompleteOrFailed(const std::vector<Future<>>& futures);
```
nit: maybe `AllFinished`?

aocsa replied:
this part of the code is more PR https://github.com/apache/arrow/pull/11017 which I needed to duplicate here in order to solve some race conditions here.

lidavidm commented on cpp/src/arrow/util/future.h:
```diff
@@ -840,6 +840,18 @@ inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s
+///
+/// The future will be marked complete if all `futures` complete
+/// successfully. Otherwise, it will be marked failed with the status of
+/// the first failing future.
```
nit: to be clear, this could be "…status of a failing future" (since "first failing future" above implies the first one to fail, not the first one in the list).

aocsa replied:
same here

lidavidm commented on cpp/src/arrow/compute/exec/sink_node.cc:
```diff
@@ -175,9 +175,9 @@ struct OrderBySinkNode final : public SinkNode {
-      bool cancelled = input_counter_.Cancel();
-      DCHECK(cancelled);
-      finished_.MarkFinished(maybe_batch.status());
+      if (input_counter_.Cancel()) {
```
Hmm, is it possible for input_counter_.Cancel() to be false?

aocsa replied:
Looks like is possible. Some runs in stress mode (`stress -m 16`) enable `DCHECK(cancelled);`. So a checks is needed.

lidavidm commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -61,9 +62,16 @@ class FilterNode : public ExecNode {
     }
-
+    std::unique_ptr<ExecNodeRunnerImpl> impl;
+    if (plan->exec_context()->executor() == nullptr) {
```
This check could perhaps just be folded into a `MakeSimpleRunner`.

lidavidm commented on cpp/src/arrow/compute/exec/filter_node.cc:
```diff
@@ -98,12 +106,23 @@ class FilterNode : public ExecNode {
+      return Status::OK();
+    };
+    auto status = impl_->SubmitTask(task);
+    if (!status.ok()) {
```
Why not fold most of this logic into the impl itself?

```cpp
if (!impl_->SubmitTask(task)) {
  StopProducing();
}
```

aocsa replied:
I addressed these comments in the refactored MapNode Class

lidavidm commented on cpp/src/arrow/compute/exec/exec_node_runner_impl.h:
```diff
@@ -0,0 +1,63 @@
+namespace arrow {
+namespace compute {
+
+class ExecNodeRunnerImpl {
```
Looking at the implementations, I wonder if this shouldn't actually just implement ExecNode? The reason being that most of the boilerplate we want to get rid of is still there right now, and additionally, there's more opportunity for implementer mistakes since you have to remember to call these new methods on the impl, and since the impl class hides the implementation, it's hard to tell at a glance if things are correct anymore. A subclass approach may not compose as well, but I'm not sure there's composition to be had in the first place here, and it would actually remove most of the boilerplate.

westonpace commented on cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:
```diff
@@ -0,0 +1,149 @@
+
+  void MarkFinished(Status status = Status::OK()) override {
+    if (!status.ok()) {
+      this->StopProducing();
+      if (input_counter_.Cancel()) {
+        this->MarkFinished();
+      }
```
`this->StopProducing()` is going to call `if (input_counter_.Cancel())` anyways so you can skip this second check.

westonpace commented on cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:
```diff
@@ -0,0 +1,149 @@
+      return;
+    }
+    task_group_.End().AddCallback([this](const Status& status) {
+      if (!status.ok()) {
+        this->finished_.MarkFinished();
+      } else if (!this->finished_.is_finished()) {
+        this->finished_.MarkFinished(status);
+      }
```
`this->finished_.MarkFinished(Status::OK())` is synonymous with `this->finished_.MarkFinished()` I think (I might be wrong on this) so you could simplify this to...

```
if (!this->finished_.is_finished()) {
  this->finished_.MarkFinished(status);
}
```

westonpace commented on cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:
```diff
@@ -0,0 +1,149 @@
+    });
+  }
+
+  void StopProducing() override {
+    this->stop_source_.RequestStop();
+    if (input_counter_.Cancel()) {
+      this->MarkFinished();
+    }
+  }
```
`RequestStop` won't take hold immediately, it simply tells the task group to stop running any additional tasks but it won't/can't interrupt any currently running tasks.  If we mark finished right away here then there is a risk that you discard the exec plan while tasks are still running.  I think, even in error scenarios, we must wait for the task group to finish before marking `finished_` complete.

aocsa replied:
`MapNode::MarkFinished()` does not mark finished right away. First It check if it is running in async mode or not, if it is async (has an executor) then it adds a callback  to the `task_group`. Maybe the name is not clear for the porpuse of this logic

westonpace reviewed (commented):
Some thoughts in addition to David's suggestion of a base class.  I see the challenge of the composition approach better now.  Since the runner doesn't have visibility of the inputs/outputs it's difficult for the runner to take care of things like forwarding signals backwards and forwards.

aocsa commented on cpp/src/arrow/compute/exec/exec_node_runner_impl.cc:
```diff
@@ -0,0 +1,149 @@
+  explicit SimpleSyncRunnerImpl(ExecContext* ctx) : ctx_(ctx) {}
+
+  Status SubmitTask(std::function<Status()> task) override {
+    if (finished_.is_finished()) {
```
it is not correct, this is just a check to not run when finished_ is already marked as Finished

aocsa commented on cpp/src/arrow/compute/exec/exec_plan.h:
```diff
@@ -243,6 +248,110 @@ class ARROW_EXPORT ExecNode {
+          if (this->input_counter_.Increment()) {
+            this->Finish(status);
+          }
+          return status;
```
Yes this is right. 
Not OK status cases are handel at the end of method `SubmitTask`.

aocsa commented:
Thanks @weston, I rebased this PR and addressed latest feedback. Moreover I ran some benchmarks to see the impact of: 1. the possible issue with ExecBatch copies; 2. async mode execution.  

Note: I runned the benchmark with the [following code ](https://github.com/apache/arrow/blob/3136a9babc8ba5fba55d35b88c7e5a967d4c01e8/cpp/src/arrow/dataset/scanner_benchmark.cc) and with this machine configuration:

```
Run on (16 X 3600 MHz CPU s) with 32Gb RAM
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 512 KiB (x8)
  L3 Unified 16384 KiB (x2)
```

1.1 Sync mode with the lambda task function capturing [batches by copy](https://github.com/apache/arrow/blob/85af59892b83fb49af58c2919d98853b9c1779fd/cpp/src/arrow/compute/exec/exec_plan.h#L311)
```
Benchmark                                               Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/min_time:1.000           3.46 ms         1.71 ms          812 items_per_second=586.126/s
MinimalEndToEndBench/1000/100/min_time:1.000         52.0 ms         38.0 ms           36 items_per_second=26.3494/s
MinimalEndToEndBench/10000/100/min_time:1.000        1102 ms          997 ms            2 items_per_second=1.00278/s
MinimalEndToEndBench/10000/1000/min_time:1.000       4752 ms         4644 ms            1 items_per_second=0.215319/s
```
1.2 Sync mode with the lambda task function capturing batches by move [with std::bind ](https://gist.github.com/aocsa/9da1f32ae1c36c133316e32d84711bc3) 
```
---------------------------------------------------------------------------------------------------------
Benchmark                                               Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/min_time:1.000           3.02 ms         1.62 ms          885 items_per_second=617.506/s
MinimalEndToEndBench/1000/100/min_time:1.000         51.7 ms         37.9 ms           35 items_per_second=26.4141/s
MinimalEndToEndBench/10000/100/min_time:1.000        1132 ms         1041 ms            1 items_per_second=0.961052/s
MinimalEndToEndBench/10000/1000/min_time:1.000       4795 ms         4680 ms            1 items_per_second=0.213687/s
```
 
As you can see ExecBatch copies were not the culprit of worst performance in some queries because there aren't "that" many ExecBatch instances and the copy is cheap (most of the case just references). 

2.  Execution without/with ThreadPool
```
without threadpool
------------------------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------
MinimalEndToEndBench/100/10/0/min_time:1.000            2.33 ms         2.33 ms          601 items_per_second=428.441/s
MinimalEndToEndBench/1000/100/0/min_time:1.000          46.8 ms         46.8 ms           30 items_per_second=21.3571/s
MinimalEndToEndBench/10000/100/0/min_time:1.000         1172 ms         1172 ms            1 items_per_second=0.853482/s
MinimalEndToEndBench/10000/1000/0/min_time:1.000        4906 ms         4905 ms            1 items_per_second=0.203876/s
MinimalEndToEndBench/10000/10000/0/min_time:1.000      52141 ms        52129 ms            1 items_per_second=0.0191832/s

with threadpool
MinimalEndToEndBench/100/10/1/min_time:1.000            3.87 ms         1.87 ms          745 items_per_second=533.584/s
MinimalEndToEndBench/1000/100/1/min_time:1.000          54.1 ms         38.3 ms           37 items_per_second=26.09/s
MinimalEndToEndBench/10000/100/1/min_time:1.000         1153 ms         1010 ms            1 items_per_second=0.990056/s
MinimalEndToEndBench/10000/1000/1/min_time:1.000        4771 ms         4624 ms            1 items_per_second=0.216249/s
MinimalEndToEndBench/10000/10000/1/min_time:1.000      49761 ms        49578 ms            1 items_per_second=0.0201702/s
```
As you can see performance when the workload is small is worst in async mode but when the workload is huge the performance tends to be better.
 

Looking forward your thoughts. cc @felipeblazing
